photon-ml/src/main/scala/com/linkedin/photon/ml/avro/model/ModelProcessingUtils.scala

```diff
@@ -174,6 +178,69 @@ object ModelProcessingUtils {
+   */
+  protected[ml] def extractGLMFeatures(glm: GeneralizedLinearModel, featureIndex: IndexMap): Array[(String, Double)] = {
+
+    val coefficients: Iterator[(Int, Double)] = glm.coefficients.means match { // (index, value)
+      case (vector: DenseVector[Double]) => vector.iterator
+      case (vector: SparseVector[Double]) => vector.activeIterator // activeIterator to iterate over the non-zeros
+    }
+    coefficients // flatMap filters out None values that can result from the case statement
+      .flatMap { case (index, value) => featureIndex.getFeatureName(index).map((_, value)) }
+      .toArray // (name, value)
+  }
+
+  /**
+   * For a given Game model and feature indexes, returns names and values for all the non-zero features in the
+   * model.
+   *
+   * Since Game models can be very large, the processing is distributed. In the case of random effects in
+   * particular, "mapPartitions" will serialize and send the content of the "iter => ..." lambda to the executors
+   * in the Spark cluster. If we didn't do that, "indexMapForRDD" would be called inside a single executor, which
+   * would result in a gigantic feature index being created inside that executor. In the case of fixed effect
+   * models (usually there is only one), we assume that the fixed effect model has a feature index that
+   * can fit within a single executor.
+   *
+   * In the case of a RandomEffectModel, there is the name e.g. "per-user" as RandomEffectModel, then
+   * there is a (sub-)model for each e.g. user, with a name such as "user 123".
+   *
+   * @param gameModel The Game model to extract all features from
+   * @param featureIndexLoaders The feature index loaders to use (each model has a separate feature index)
+   * @return All the (feature name, feature value) pairs of that Game model
+   *         The first Map's keys are like "fixed", "per-user". Then the RDD String keys are like
+   *         "user 123". Finally, for that user, the RDD element contains an array of pairs
+   *         (feature name, feature value).
+   */
+  protected[ml] def extractGameModelFeatures(
+    sc: SparkContext,
+    gameModel: GAMEModel,
+    featureIndexLoaders: Map[String, IndexMapLoader]): Map[String, RDD[(String, Array[(String, Double)])]] = {
+
+    gameModel.toMap.map {
+
+      case (fixedEffectName: String, model: FixedEffectModel) =>  // e.g. fixedEffectName = "fixed"
+        val featureIndex = featureIndexLoaders(fixedEffectName).indexMapForRDD()
+        (fixedEffectName, sc.parallelize(Array((fixedEffectName, extractGLMFeatures(model.model, featureIndex)))))
+
+      case (randomEffectName: String, model: RandomEffectModel) => // e.g. randomEffectName = "per-user"
+        (randomEffectName, model.modelsRDD.mapPartitions { iter =>
+          iter.map { case (subModelId, subModel) => // e.g. subModelId = "user 123"
+            (subModelId, extractGLMFeatures(subModel, featureIndexLoaders(subModelId).indexMapForRDD())) }
```

Comment:
We need to call `indexMapForRDD` before entering this closure. Otherwise, it's going to try loading the index map for each datapoint.